### DISCUSSION TITLE
HELP: How to use Complexheatmap or Oncoprint libraries for Non-Medical usecase

### DISCUSSION THREAD

**Discussion by ekrishgit** [jokergoo/ComplexHeatmap#991] [Q&A]
First of all I want to say that am Impressed by ComplexHeatmap/Oncoprint visualizations and I would like to take this inspiration and apply to non-Medical use case. For ex., I have lot of data processing jobs that runs on different schedules and goes thru set of control checks. I want to produce a visual to show the jobs, processes and their statuses in below format. My existing application is developed on React JS/Java. What library can I use to yield similar visual results. Any help or direction is much appreciated!!

Below is how I intent to replace the values

Patient ID - Job ID
Metastatic site - Run Instance (1, 2, 3 etc.,)
Gene Panel - Process Name
Genetic Alterations - Control checks

![image](https://user-images.githubusercontent.com/42822460/191621023-b81c9438-041d-4809-a4ff-43b21fd57b20.png)

**Comment by jokergoo** [comment]
If you can share a piece of your data, I'd like to help.

**Comment by ekrishgit** [comment]
Thank you! Below is sample data in pipe delimited format. First row is header.

Number of Process - 1 .. N
Number of SubProcess - 1 .. N
Control - Limited options. In the example 3 Options. Would like to plot only failures.
Number of Dates - 1 .. End of Month

Date | Process | SubProcess | Control | Result
20220101 | A | 1 | No control |  
20220101 | A | 2 | Mandatory | Fail
20220101 | A | 3 | No control |  
20220101 | A | 4 | No control |  
20220102 | A | 1 | Mandatory | Fail
20220102 | A | 2 | Optional | Fail
20220102 | A | 3 | No control |  
20220102 | A | 4 | No control |  
20220102 | A | 4 | Optional | Fail
20220101 | B | 1 | No control |  
20220101 | B | 2 | No control |  
20220101 | B | 3 | No control |  
20220102 | B | 1 | Mandatory | Fail
20220102 | B | 2 | Optional | Fail
20220102 | B | 3 | Mandatory | Fail

Would like to plot above data as below with legend.

![image](https://user-images.githubusercontent.com/42822460/193356279-cc44f62e-8df3-4aaa-8a9d-e3513cc8548f.png)

**Comment by jokergoo** [comment]
you need to construct a matrix where rows are date and columns are combination of "process"
and "subprocess". It should not be difficult to convert your original data. Here I just generate 
a random one:


```r
mat = matrix(
    sample(c("No control", "Mandatory", "Optional", ""), size = 100, p = c(1, 1, 1, 4), replace = TRUE),
	nrow = 10, ncol = 10)
rownames(mat) = paste0("2022-1-", 1:10)
colnames(mat) = c(paste0("A", 1:5), paste0("B", 1:5))
mat
```

```
          A1           A2          A3          A4           A5
2022-1-1  ""           ""          "Optional"  ""           "Mandatory"
2022-1-2  "Mandatory"  ""          ""          "Optional"   "Mandatory"
2022-1-3  "No control" ""          ""          ""           "Optional"
2022-1-4  "Optional"   ""          ""          ""           ""
2022-1-5  ""           "Optional"  "Optional"  ""           ""
2022-1-6  ""           "Optional"  ""          ""           ""
2022-1-7  ""           "Mandatory" ""          ""           "Optional"
2022-1-8  "Mandatory"  ""          "Mandatory" "No control" ""
2022-1-9  ""           ""          "Mandatory" "Mandatory"  "Mandatory"
2022-1-10 ""           "Mandatory" ""          "Optional"   "Mandatory"
          B1           B2           B3          B4           B5
2022-1-1  ""           ""           "Mandatory" ""           "Mandatory"
2022-1-2  ""           ""           "Optional"  ""           ""
2022-1-3  ""           "Mandatory"  ""          ""           "Optional"
2022-1-4  ""           "Optional"   "Mandatory" ""           "Mandatory"
2022-1-5  "Mandatory"  ""           ""          ""           "Optional"
2022-1-6  ""           "No control" ""          "No control" "No control"
2022-1-7  "No control" ""           "Mandatory" ""           "Optional"
2022-1-8  ""           ""           ""          ""           "Mandatory"
2022-1-9  ""           "Mandatory"  "Optional"  ""           "Optional"
2022-1-10 "Mandatory"  ""           ""          "No control" ""
```

Then in the heatmap, "process" and "subprocess" can be two column annotations.
And the graphics for "Control" can be implemented with a `cell_fun`:

```r
control_col = c("Mandatory" = "red", "Optional" = "yellow")
Heatmap(mat, show_heatmap_legend = FALSE,
	rect_gp = gpar(type = "none"),
	top_annotation = HeatmapAnnotation(process = c(rep("A", 5), rep("B", 5)),
    	subprocess = c(1:5, 1:5),
    	col = list(subprocess = c("1" = "red", "2" = "orange", "3" = "blue", "4" = "green", "5" = "purple"))
    ),
    cell_fun = function(j, i, x, y, w, h, f) {
    	grid.rect(x, y, w, h, gp = gpar(fill = "grey", col = "black"))
    	if(mat[i, j] %in% names(control_col)) {
    		grid.points(x, y, pch = 16, gp = gpar(col = control_col[mat[i, j]]))
    	}
    })
```

<img width="845" alt="image" src="https://user-images.githubusercontent.com/449218/193805197-5d394941-0e16-46a7-ba05-61bd86ce8261.png">

**Reply by ekrishgit** [reply]
Thank you for the pointers!! I am applying the logic and implementing using React Heatmap.
